### ISSUE TITLE
Predict local_property

### ISSUE THREAD

**Issue by tcnicholas** [libAtoms/QUIP#565]
Dear GAP/QUIP developers,

I have fit a gap model to predict a local property (local_property_parameter_name=myprop), and used a per-atom SOAP descriptor. Once I have my xml files, it is not clear to me how to then e.g. use quippy to predict the local property for a given ASE Atoms object?

Any help/suggestions would be appreciated.
Thomas

**Comment by albapa** [comment]
I don't have such an XML file handy but I think it's as simple as predicting the local energy. With command line `quip` you'd just add `local` to the command line. Of course the "energy" property doesn't make much sense, since that adds all those terms.

**Comment by bernstei** [comment]
I think @tcnicholas is asking for a python interface.  Depends on whether the use of "quippy" and reference to "ASE atoms object" were literal .

**Comment by tcnicholas** [comment]
Yes, I was wondering if there was a Python interface. I.e., set-up the quippy Potential object and then calculating the specific local property (by analogy to e.g. `gap_pot.get_forces(atoms)`)

```
from quippy.potential import Potential

gap_pot = Potential(param_filename='gap.xml')
gap_pot.get_local_property(atoms)
```

**Comment by albapa** [comment]
Would `gap_pot.get_potential_energies(atoms)` work?

**Comment by bernstei** [comment]
Wouldn't that return the energies, not the local property?  My current solution is to avoid the whole messy ASE `Atoms.get_[potential_energy/forces/stress]` for non-standard properties (since that interface is inherently inflexible), and just have the calculator store non-standard results in `Calculator.results_extra["some_key"]`, analogous to `Calculator.results` for standard properties.  We could have `Potential` store the results under `Calculator.results_extra["local_property"]`

**Comment by albapa** [comment]
In this case the two coincide. @tcnicholas used a hypothetical local energy property to fit a GAP, so the prediction will return the same quantity as the local energy.

**Comment by mcaroba** [comment]
The way how we handle per-atoms properties in TurboGAP for local energies and Hirshfeld volumes (when available) is through the array construction, which is an Atoms property, not a Calculator property (in that sense, perhaps more "portable"?). You access that simply by doing atoms.get_array("local_property_name") and likely for adding this property you do atoms.set_array("local_property_name", numpy_array_with_local_property). When a specific atom has no local property associated with it this, the corresponding array element can be replaced by np.NaN or something like that.

**Comment by tcnicholas** [comment]
After a couple of experiments, @albapa is right. The `gap_pot.get_potential_energies(atoms)` does just equate to the per-atom local property prediction. (Note, I needed to make sure the e0 values for were set to zero.)

**Comment by bernstei** [comment]
So `local_property_param_name` fits a local energy, i.e. the resulting model returns the values as the calculated energy?  I would have expected, in that case, to have named it `local_energy_param_name`.  Or is it a distinct local property in the fortran code, and it's only the quippy ASE interface that returns it in place of energy?
